### PR TITLE
delete stale build DB schemas

### LINKED ISSUE #307: draft build schemas in DB shouldn't persist forever
## problem
- the DB storage taken up by draft builds is increasing with no mechanism to delete old drafts
- currently using 40% of the 280 GB storage

## background
- we now use our persistent DB to build multiple data products (TDB, Pluto, DevDB, CPDB, KPDB, ZAP) and intend to do the same for all of them
- builds use a schema (a collection of tables) named after the unique [`build_name`](https://github.com/NYCPlanning/data-engineering/blob/dm-pluto-in-edm-db/dcpy/connectors/edm/build_metadata.py#L8)
  - this schema name is identical to the S3 folder name for build outputs via [`DraftKey`](https://github.com/NYCPlanning/data-engineering/blob/dm-pluto-in-edm-db/products/template/build_scripts/export.py#L36)
- builds are either triggered manually or by PR tests
- approach ideas in https://github.com/NYCPlanning/data-engineering/issues/290 and [this issue comment](https://github.com/NYCPlanning/data-engineering/issues/17#issuecomment-1749420510)

### PR DESCRIPTION
**Pull request by damonmcc**
resolves #307 

## approach
- compile the list of active build names: all remote branch names and all open PR numbers
  - uses `build_metadata.build_name()` to go from the branch/PR to the build name
- in all build DBs, delete all schemas that aren't in the list of active builds names

## tests
[successful action run](https://github.com/NYCPlanning/data-engineering/actions/runs/6656198668/job/18088194967) deleting stale TDB schemas
- before:
  - Screenshot 2023-10-26 at 11 01 12 AM
- after:
  - image

### CODE REVIEW

**Comment by damonmcc** [comment]
before actually letting it delete schemas, wanted to get eyes on the approach @sf-dcp @alexrichey @fvankrieken 

[here's](https://github.com/NYCPlanning/data-engineering/actions/runs/6646795783/job/18060960165#step:4:8) a link to the logs of the latest run. the action is scheduled to run every Saturday at midnight via `cron: "0 0 * * SUN"`

File: dcpy/utils/postgres.py [hunk 1]
```diff
@@ -7,6 +7,11 @@
 from sqlalchemy import create_engine, text
 
 DEFAULT_POSTGRES_SCHEMA = "public"
+PROTECTED_POSTGRES_SCHEMAS = [
+    DEFAULT_POSTGRES_SCHEMA,
+    "information_schema",
+    "pg_catalog",
+]
```
Comment (damonmcc) [inline dcpy/utils/postgres.py:14]: these are schemas we'd never want to delete

File: dcpy/connectors/edm/build_metadata.py [hunk 2]
```diff
@@ -5,10 +5,18 @@
...
 
 
-def build_name() -> str:
+def build_name(event: str | None = None, branch: str | None = None) -> str:
     if os.environ.get("BUILD_ENGINE_SCHEMA"):
         return os.environ["BUILD_ENGINE_SCHEMA"]
-    return git.run_name()
+
+    event = git.event_name() if not event else event
+    branch = git.branch() if not branch else branch
```
Comment (alexrichey) [inline dcpy/connectors/edm/build_metadata.py:13]: fyi, you can just do `branch or git.branch()` (and same for `event`)

**Review by alexrichey — APPROVED** [review]
Great stuff!

**Pull request merged**